examples/gaussian_process/svgp.py

```diff
@@ -0,0 +1,216 @@
+
+        # With ill-conditioned Kzz, the inverse is often asymmetric, which
+        # breaks further cholesky decomposition. We compute a symmetric one.
+        Kzz_chol_inv = tf.matrix_triangular_solve(
+            Kzz_chol, tf.eye(hps.N_Z, dtype=hps.dtype))
+        Kzz_inv = tf.matmul(tf.transpose(Kzz_chol_inv), Kzz_chol_inv)
+
+        Kxziz = tf.matmul(Kxz, Kzz_inv)
+        mean_y = tf.matmul(fZ, tf.transpose(Kxziz))  # [N_particles, N_X]
+
+        cov_xz = rbf(X_ph, X_ph, k_inv_scale) - \
+            tf.matmul(Kxziz, tf.transpose(Kxz))
+        cov_y = cov_xz + noise_level * tf.eye(
+            tf.shape(X_ph)[0], dtype=hps.dtype)
+        if full_cov:
+            cov_y_chol = tf.tile(tf.expand_dims(tf.cholesky(cov_y), 0),
+                                 [n_particles, 1, 1])
+            Y = zs.MultivariateNormalCholesky('Y', mean_y, cov_y_chol)
+        else:
+            std_y = tf.sqrt(tf.matrix_diag_part(cov_y))
+            std_y = tf.tile(tf.expand_dims(std_y, 0), [n_particles, 1])
+            Y = zs.Normal('Y', mean=mean_y, std=std_y, group_ndims=1)
+
+    return model, Y
+
+
+def build_variational(hps, n_particles):
+    with zs.BayesianNet() as variational:
+        Z_mean = tf.get_variable(
+            'Z/mean', [hps.N_Z], hps.dtype, tf.zeros_initializer())
+        Z_cov_raw = tf.get_variable(
+            'Z/cov_raw', initializer=tf.eye(hps.N_Z, dtype=hps.dtype))
+        Z_cov_tril = tf.matrix_set_diag(
+            tf.matrix_band_part(Z_cov_raw, -1, 0),
+            tf.nn.softplus(tf.matrix_diag_part(Z_cov_raw)))
+        Z = zs.MultivariateNormalCholesky(
+            'fZ', Z_mean, Z_cov_tril, n_samples=n_particles)
+    return variational
+
+
+def main():
+    tf.set_random_seed(1237)
+    np.random.seed(1234)
+    hps = parser.parse_args()
+
+    # Load data
+    data_path = os.path.join(conf.data_dir, hps.dataset + '.data')
+    data_func = getattr(dataset, 'load_uci_' + hps.dataset)
+    x_train, y_train, x_valid, y_valid, x_test, y_test = data_func(data_path)
+    x_train = np.vstack([x_train, x_valid])
+    y_train = np.hstack([y_train, y_valid])
+    N, hps.N_X = x_train.shape
+    hps.dtype = getattr(tf, hps.dtype)
+
+    # Standardize data
+    x_train, x_test, _, _ = dataset.standardize(x_train, x_test)
+    y_train, y_test, mean_y_train, std_y_train = dataset.standardize(
+        y_train, y_test)
+
+    # Build graph
+    X_ph = tf.placeholder(hps.dtype, [None, hps.N_X], 'X')
+    Y_ph = tf.placeholder(hps.dtype, [None], 'Y')
+    Z_pos = tf.get_variable('Z/pos', [hps.N_Z, hps.N_X], hps.dtype,
+                            initializer=tf.random_uniform_initializer(-1, 1))
+    N_particles = tf.placeholder(tf.int32, [], 'n_particles')
+
+    # Train
+    def log_joint(observed):
+        model, _ = build_model(observed, Z_pos, X_ph, hps, N_particles)
+        prior, ll = model.local_log_prob(['fZ', 'Y'])
+        n_batch = tf.cast(tf.shape(X_ph)[0], hps.dtype)
+        return prior + ll / n_batch * N
```

Comment:
no need to divide by `n_batch` since you do `reduce_mean` outside.

Reply:
Note the `log_prob` for `Y` is summed over the whole batch [here](https://github.com/meta-inf/zhusuan/blob/00105c7d7f856821d38f334ce9b3f129c544abf9/examples/gaussian_process/svgp.py#L98). The `reduce_mean` is for variational posterior particles.

Reply:
In that sense there is no need for `tf.reduce_mean(lower_bound)` as well as `tf.reduce_mean(cost)`?

Reply:
Right, thanks.